### PR TITLE
Feature/update to godot 3.2

### LINKED ISSUE #1: Update to godot 3.2 [Moved from old Repo]
Describe the problem or limitation you are having in your project:
The used godot headers api json is quite old. It has to be updated to ensure compability with current and future godot versions.

Describe how this feature / enhancement will help you overcome this problem or limitation:
No limitations just yet. But there will be if we don't keep this up to date.

Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:
None to provide.

Describe implementation detail for your proposal (in code), if possible:
Pull the latest changes from the master branch of the godot headers api (submodule in this project).
Ensure compability.

If this enhancement will not be used often, can it be worked around with a few lines of code?:
No.

Is there a reason why this should be in this project and not individually solved?:
Yes otherwise this project will not work in the future.

### LINKED ISSUE #9: Use receivers for the property lambdas [Moved from old repo]
**Describe the problem or limitation you are having in your project:**
Local value modification work-around is somewhat annoying.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
By changing the dsl-like interface like `position { it.x = 1 }` to using a receiver instead, the code will look like `position { x = 1 }` instead. Ie. not require the `it`

Not sure it this has ever been discussed or if it's a good solution.

Maybe it is possible to circumvent the problem by using delegates, or heck, maybe one doesn't want to circumvent the problem at all.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**

**Describe implementation detail for your proposal (in code), if possible:**
```diff
diff --git a/tools/api-classes-generator/src/main/kotlin/Class.kt b/tools/api-classes-generator/src/main/kotlin/Class.kt
index 681cf69..501f204 100644
--- a/tools/api-classes-generator/src/main/kotlin/Class.kt
+++ b/tools/api-classes-generator/src/main/kotlin/Class.kt
@@ -231,7 +231,7 @@ class Class(
                                     ParameterSpec.builder(
                                             "schedule",
                                             LambdaTypeName.get(
-                                                    parameters = *arrayOf(parameterTypeName),
+                                                    receiver = parameterTypeName,
                                                     returnType = ClassName("kotlin", "Unit")
                                             )
                                     ).build()
```
**If this enhancement will not be used often, can it be worked around with a few lines of code?:**

**Is there a reason why this should be in this project and not individually solved?:**
The code is generated by the annotation processor.

### PR DESCRIPTION
**Pull request by piiertho**
This updates bindings for use with godot 3.2.
Update GDNative headers to 3.2
Change core classes for godot 3.2

It seems  memscope mentionned in #1  are not mandatory as we're not using pointers in methods that don't have them.
Methods with no argument could be transformed into a backing field, like `Vector3.normalize`, but this is not the subject of this PR.
Also, we should wait #8 is merged in this branch before merging this one in master.

Please guys review this PR carrefully, as core classes has changed.

This resolves #1 and resolves #9

### CODE REVIEW

**Comment by chippmann** [comment]
@piiertho Should we still wait? As @lassem wrote here https://github.com/utopia-rise/godot-kotlin/pull/8#issuecomment-583617356 he would rebase from master and then merge it into master which is also fine for me.

File: samples/games/project/samples.gdnlib
Comment (chippmann) [inline samples/games/project/samples.gdnlib]: Shouldn't we just remove the 32 bit? As neither modern macOS versions nor godot 3.2 support 32bit on macOS anymore.
Additionally we can't even build this project on 32bit systems because of to less memory available on them.

File: wrapper/godot-library/src/main/kotlin/godot/core/Color.kt
Comment (chippmann) [inline wrapper/godot-library/src/main/kotlin/godot/core/Color.kt:8]: Just as a note:
Personally I think in the feature we should go with the scout boys approach and leave to code cleaner as we found it. Means: From our predecessors there is some pretty ugly formatting in all classes. So my suggestion is to reformat the whole class when we need to modify anything in it. With this, all the classes will get clean and concise code over time.
What do you think @piiertho ? Just a suggestion that came to my mind looking at this class in general :-) Also you would not have to do it in this pr as it's too important for now :-)

**Comment by lassem** [comment]
I have rebased #8 on top of this branch, so whatever you decide is fine with me.
Merging this first, then having me rebase on master as a separate PR is probably the cleanest.

File: samples/games/kotlin/build.gradle.kts
Comment (lassem) [inline samples/games/kotlin/build.gradle.kts:71]: Isn't dropping the named lambda argument a step in the wrong direction?

File: samples/games/kotlin/src/main/kotlin/godot/samples/games/shmup/EnemyPath.kt
Comment (lassem) [inline samples/games/kotlin/src/main/kotlin/godot/samples/games/shmup/EnemyPath.kt:15]: Seems like someone has gone through some effort to keep this inside an `apply` block.
We could put `addChild(this)` and `start()` inside here as well. No biggie though.

**Review by chippmann — CHANGES_REQUESTED** [review]
Very great work! Thanks a lot!
One thing i noticed that in my opinion needs changing (my bad I didn't saw this in the rewriting to kotlinpoet):
The klaxon lib is still at version 3.0.1 (current 5.2) which has dependencies on kotlin version 1.2.10
see: `./gradlew :tools:api-classes-generator:dependencies`
I think this needs fixing.

**Review by lassem — COMMENTED** [review]
LGTM. However there's a lot of stuff in here that is new to me so. Only nit-picking this time 😂

**Comment by chippmann** [comment]
> I have rebased #8 on top of this branch, so whatever you decide is fine with me.
> Merging this first, then having me rebase on master as a separate PR is probably the cleanest.

Ok cool i will mergi it now then :-)

File: samples/games/kotlin/src/main/kotlin/godot/samples/games/shmup/EnemyPath.kt
Comment (chippmann) [reply samples/games/kotlin/src/main/kotlin/godot/samples/games/shmup/EnemyPath.kt:15]: I thought of that as well but I think the `addChild` would be out of context here and has nothing to do with `apply` which in my opinion suggests that the block is applying something to the tween.

File: samples/games/kotlin/build.gradle.kts
Comment (chippmann) [reply samples/games/kotlin/build.gradle.kts:71]: I think not as `it.compilations.getByName("main")` also has a `target` so in it's block it's hard to understand which `target` we mean.
But i pushed a change which should clarify this by adding an explicit `this` before the usages. Feel free to look at that and say what you think about it :-)

File: samples/games/kotlin/src/main/kotlin/godot/samples/games/shmup/EnemyPath.kt
Comment (lassem) [reply samples/games/kotlin/src/main/kotlin/godot/samples/games/shmup/EnemyPath.kt:15]: True!

Comment (piiertho) [reply samples/games/kotlin/src/main/kotlin/godot/samples/games/shmup/EnemyPath.kt:15]: Yes call start here was creating a bug, causing the tween to not start. It comes from Godot 3.1 if I remember well.

File: samples/games/project/samples.gdnlib
Comment (piiertho) [reply samples/games/project/samples.gdnlib]: Solved by b1371cc :)

**Comment by piiertho** [comment]
Trying to update Klaxon introduce a bug in api classes generation:
```
Exception in thread "main" java.lang.IllegalArgumentException: Can't escape identifier `feedback/active` because it contains illegal characters: /
	at com.squareup.kotlinpoet.UtilKt.failIfEscapeInvalid(Util.kt:204)
	at com.squareup.kotlinpoet.UtilKt.escapeIfNecessary(Util.kt:212)
	at com.squareup.kotlinpoet.UtilKt.escapeIfNecessary$default(Util.kt:209)
	at com.squareup.kotlinpoet.CodeBlock$Builder.addArgument(CodeBlock.kt:341)
	at com.squareup.kotlinpoet.CodeBlock$Builder.add(CodeBlock.kt:317)
	at com.squareup.kotlinpoet.CodeBlock$Companion.of(CodeBlock.kt:447)
	at com.squareup.kotlinpoet.CodeWriter.emitCode(CodeWriter.kt:214)
	at com.squareup.kotlinpoet.PropertySpec.emit$kotlinpoet(PropertySpec.kt:81)
	at com.squareup.kotlinpoet.PropertySpec.emit$kotlinpoet$default(PropertySpec.kt:59)
	at com.squareup.kotlinpoet.TypeSpec.emit$kotlinpoet(TypeSpec.kt:253)
	at com.squareup.kotlinpoet.TypeSpec.emit$kotlinpoet$default(TypeSpec.kt:96)
	at com.squareup.kotlinpoet.FileSpec.emit(FileSpec.kt:162)
	at com.squareup.kotlinpoet.FileSpec.writeTo(FileSpec.kt:63)
	at com.squareup.kotlinpoet.FileSpec.writeTo(FileSpec.kt:91)
	at com.squareup.kotlinpoet.FileSpec.writeTo(FileSpec.kt:96)
	at Class.generate(Class.kt:75)
	at GeneratorKt.main(Generator.kt:16)
	at GeneratorKt.main(Generator.kt)
```

This is because it does not replace / by _ anymore like for this property : `feedback/active` in `AudioEffectDelay` (in api.json from godot_headers).

So I think updating this lib should not be done in this branch.
Also, code is working for now, so updating this library should be low prio, don't forget we have to support android and ios :P
But I agree, we should find a workaround to polish project.

**Comment by chippmann** [comment]
> Trying to update Klaxon introduce a bug in api classes generation:
> 
> ```
> Exception in thread "main" java.lang.IllegalArgumentException: Can't escape identifier `feedback/active` because it contains illegal characters: /
> 	at com.squareup.kotlinpoet.UtilKt.failIfEscapeInvalid(Util.kt:204)
> 	at com.squareup.kotlinpoet.UtilKt.escapeIfNecessary(Util.kt:212)
> 	at com.squareup.kotlinpoet.UtilKt.escapeIfNecessary$default(Util.kt:209)
> 	at com.squareup.kotlinpoet.CodeBlock$Builder.addArgument(CodeBlock.kt:341)
> 	at com.squareup.kotlinpoet.CodeBlock$Builder.add(CodeBlock.kt:317)
> 	at com.squareup.kotlinpoet.CodeBlock$Companion.of(CodeBlock.kt:447)
> 	at com.squareup.kotlinpoet.CodeWriter.emitCode(CodeWriter.kt:214)
> 	at com.squareup.kotlinpoet.PropertySpec.emit$kotlinpoet(PropertySpec.kt:81)
> 	at com.squareup.kotlinpoet.PropertySpec.emit$kotlinpoet$default(PropertySpec.kt:59)
> 	at com.squareup.kotlinpoet.TypeSpec.emit$kotlinpoet(TypeSpec.kt:253)
> 	at com.squareup.kotlinpoet.TypeSpec.emit$kotlinpoet$default(TypeSpec.kt:96)
> 	at com.squareup.kotlinpoet.FileSpec.emit(FileSpec.kt:162)
> 	at com.squareup.kotlinpoet.FileSpec.writeTo(FileSpec.kt:63)
> 	at com.squareup.kotlinpoet.FileSpec.writeTo(FileSpec.kt:91)
> 	at com.squareup.kotlinpoet.FileSpec.writeTo(FileSpec.kt:96)
> 	at Class.generate(Class.kt:75)
> 	at GeneratorKt.main(Generator.kt:16)
> 	at GeneratorKt.main(Generator.kt)
> ```
> 
> This is because it does not replace / by _ anymore like for this property : `feedback/active` in `AudioEffectDelay` (in api.json from godot_headers).
> 
> So I think updating this lib should not be done in this branch.
> Also, code is working for now, so updating this library should be low prio, don't forget we have to support android and ios :P
> But I agree, we should find a workaround to polish project.

True. I'm fine with upgrading it later.

File: samples/games/kotlin/build.gradle.kts
Comment (chippmann) [reply samples/games/kotlin/build.gradle.kts:71]: @lassem Ok for you if i leave it like this?

**Review by chippmann — APPROVED** [review]
Good work!
After this change i think we can make a minor bump and release.
The discussion regarding versioning can be delayed until the next major bump. Ok @piiertho ?

**Comment by piiertho** [comment]
Yes after merging we can up version and make a release.
I'm just waiting for @CedNaru for cinterop code review :)

**Review by CedNaru — APPROVED** [review]

**Pull request merged**
